v2/blockstore/readwrite.go

```diff
@@ -59,50 +66,164 @@ func WithIndexPadding(p uint64) Option {
-		return nil, err
-	}
 
-	v1Header := &carv1.CarHeader{
-		Roots:   roots,
-		Version: 1,
+	fFlag := os.O_RDWR | os.O_CREATE
+	if !b.resume {
+		fFlag = fFlag | os.O_EXCL
 	}
-	if err := carv1.WriteHeader(v1Header, b.carV1Writer); err != nil {
-		return nil, fmt.Errorf("couldn't write car header: %w", err)
+	var err error
+	b.f, err = os.OpenFile(path, fFlag, 0o666) // TODO: Should the user be able to configure FileMode permissions?
+	if err != nil {
+		return nil, fmt.Errorf("could not open read/write file: %w", err)
 	}
+	b.carV1Writer = internalio.NewOffsetWriter(b.f, int64(b.header.CarV1Offset))
+	v1r := internalio.NewOffsetReadSeeker(b.f, int64(b.header.CarV1Offset))
+	b.ReadOnly = ReadOnly{backing: v1r, idx: b.idx, carv2Closer: b.f}
+
+	if b.resume {
+		if err = b.resumeWithRoots(roots); err != nil {
+			return nil, err
+		}
+	} else {
+		if err = b.initWithRoots(roots); err != nil {
+			return nil, err
+		}
+	}
+
 	return b, nil
 }
 
+func (b *ReadWrite) initWithRoots(roots []cid.Cid) error {
+	if _, err := b.f.WriteAt(carv2.Pragma, 0); err != nil {
+		return err
+	}
+	return carv1.WriteHeader(&carv1.CarHeader{Roots: roots, Version: 1}, b.carV1Writer)
+}
+
+func (b *ReadWrite) resumeWithRoots(roots []cid.Cid) error {
+	// On resumption it is expected that the CAR v2 Pragma, and the CAR v1 header is successfully written.
+	// Otherwise we cannot resume from the file.
+	// Read pragma to assert if b.f is indeed a CAR v2.
+	version, err := carv2.ReadVersion(b.f)
+	if err != nil {
+		// The file is not a valid CAR file and cannot resume from it.
+		// Or the write must have failed before pragma was written.
+		return err
+	}
+	if version != 2 {
+		// The file is not a CAR v2 and we cannot resume from it.
+		return fmt.Errorf("cannot resume on CAR file with version %v", version)
+	}
+
+	// Use the given CAR v1 padding to instantiate the CAR v1 reader on file.
+	v1r := internalio.NewOffsetReadSeeker(b.ReadOnly.backing, 0)
+	header, err := carv1.ReadHeader(bufio.NewReader(v1r))
+	if err != nil {
+		// Cannot read the CAR v1 header; the file is most likely corrupt.
+		return fmt.Errorf("error reading car header: %w", err)
+	}
+	if !header.Equals(carv1.CarHeader{Roots: roots, Version: 1}) {
+		// Cannot resume if version and root does not match.
+		return errors.New("cannot resume on file with mismatching data header")
+	}
+
+	// TODO See how we can reduce duplicate code here.
+	// The code here comes from index.Generate.
+	// Copied because we need to populate an insertindex, not a sorted index.
+	// Producing a sorted index via generate, then converting it to insertindex is not possible.
+	// Because Index interface does not expose internal records.
+	// This may be done as part of https://github.com/ipld/go-car/issues/95
+
+	offset, err := carv1.HeaderSize(header)
+	if err != nil {
+		return err
+	}
+	frameOffset := int64(0)
+	if frameOffset, err = v1r.Seek(int64(offset), io.SeekStart); err != nil {
+		return err
+	}
+
+	for {
```

Comment:
i would at least pull this utility into the index package rather than having it inline here

Reply:
That makes sense. I thought I do this as part of #118 where most of the index types need to be unexposed and moved back to internal packages. 

Doing this now would either mean exposing more things in `index` package that we need to unexpose later or making this PR much larger by also doing #118.

If you agree, I'll add a TODO here and attend to it in a separate PR.